### PR TITLE
Feature: branch protection conversation resolution

### PR DESCRIPTION
**Pull request by josh-barker**
This PR addresses #868 and updates the following resources:

- github_branch_protection
- github_branch_protection_v3

You may find it easier to review commit by commit.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -10,7 +10,7 @@ require (
 	github.com/hashicorp/hcl/v2 v2.3.0 // indirect
 	github.com/hashicorp/terraform-config-inspect v0.0.0-20191212124732-c6ae6269b9d7 // indirect
 	github.com/hashicorp/terraform-plugin-sdk v1.7.0
-	github.com/shurcooL/githubv4 v0.0.0-20201206200315-234843c633fa
+	github.com/shurcooL/githubv4 v0.0.0-20210725200734-83ba7b4c9228
```
Comment (jcudit) [inline go.mod:13]: Can we get some insight into why this dependency is changing?  Not a blocker, but hoping to understand what we're currently missing.

**Review by jcudit — COMMENTED** [review]
Looking like a promising contribution so far with the following caveats:

- Can we get some clarity on the v4 dependency update?
- Can we drop the edits to the `_v3` resource? It has been in a frozen state since released so that we do not diverge from the recommended resource.

File: go.mod (hunk 1 above)
Comment (josh-barker) [reply go.mod:13]: @jcudit The structs/variables for Conversation Resolution do not exist with the older version of `githubv4`.

**Comment by josh-barker** [comment]
@jcudit There are existing issues relating to performance of the branch protection resource using GraphQL.
For example https://github.com/integrations/terraform-provider-github/issues/606.

Therefore, I've implemented the configuration in both resources so that people have the option of choosing the more performant resource.

Personally, I can't use the `github_branch_protection` as I am managing too many repositories.

How did you want to proceed?

**Comment by jcudit** [comment]
I think we can ship to both resources in this case, but overall discourage adding more to the `github_branch_protection_v3` resource.  Aiming to send this out in an upcoming release.

**Review by jcudit — APPROVED** [review]
Tests well, thanks for the contribution!

**Comment by mambax** [comment]
Sorry but quick question. The [docs state](https://github.com/integrations/terraform-provider-github/blob/main/website/docs/r/branch_protection.html.markdown) that this new `require_conversation_resolution ` is `optional`.

Yet, after updating the provider our plans break:
```
github_team_repository.qa-microservice-streaming-api-omega-admin: Refreshing state... [id=9733:qa-microservice-streaming-api]

Error: Field 'requiresConversationResolution' doesn't exist on type 'BranchProtectionRule'

  with github_branch_protection.actions,
  on 30_actions.tf line 19, in resource "github_branch_protection" "actions":
  19: resource "github_branch_protection" "actions" {
```

Is that an issue or wanted behaviour?

**Pull request merged**